[심기성] 레이싱게임 step2 리뷰 부탁드립니다.

SimKiSeong opened the pull request:
레이싱게임 step2 리뷰 부탁드립니다.

SimKiSeong committed:
feature : first version

hyukjin-lee commented on src/main/java/com/econo/racingGame/controller/RankController.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public int bigInteger(int a, int b){
+        if(a>b){
```
제출하기 전 라인정리 해주세요, bigInteger는 함수이름으로 적당하지 않아보입니다.

hyukjin-lee commented on src/main/java/com/econo/racingGame/controller/RankController.java:
```diff
@@ -0,0 +1,41 @@
+        return NONE;
+    }
+
+    public String[] returnWinner(Car[] cars){
```
함수 이름에 return같은 문법 언어를 쓰지 않습니다.

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/GameController.java:
```diff
@@ -6,16 +6,17 @@
         for(int i = 0;i<cars.length;i++){
             cars[i] = new Car();
+            cars[i].setDriver(names[i]);
             cars[i].setMoveDistance(returnDistance(tryNumber));
```
메소드 네이밍으로 봤을 때 make만 하는 줄 알았는데, 경주까지 하고 있네요
분리하는게 좋지 않을까요?

esp2ar0 replied:
makeCars 메소드에서 많은일을 하고 있네요.
또한 아래의 메소드들도 '레이싱 게임'이라는 관점으로 네이밍을 하면 좋겠어요

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/RankController.java:
```diff
@@ -0,0 +1,41 @@
+    private final String SPLIT_MARK = ",";
+    private final String NONE = "";
+
+    public String findWinner(int farDistance,Car car){
```
findFarDistance 메소드와 이 메소드가 수행하는 로직이 많이 겹친다고 느껴집니다.
좀 더 효율적으로 처리할 수 있는 방법이 없을까요?

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/RankController.java:
```diff
@@ -0,0 +1,41 @@
+        return farDistance;
+    }
+
+    public int bigInteger(int a, int b){
```
이러한 기능을 하는 메소드가 라이브러리에 있지 않나요?

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/RankController.java:
```diff
@@ -0,0 +1,41 @@
+        String winner = new String();
+        int farDistance = findFarDistance(cars);
+        for(int i = 0;i<cars.length;i++){
+            winner = winner + findWinner(farDistance,cars[i]);
```
문자열을 합치는 방법은 어떤 게 있으며, 효율적인 방법은 무엇일까요?

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/GameController.java:
```diff
@@ -6,16 +6,17 @@
-        Car[] cars = new Car[carNumber];
+    public Car[] makeCars(String[] names,int tryNumber){
+        Car[] cars = new Car[names.length];
         for(int i = 0;i<cars.length;i++){
```
정렬이 안돼 있네요

esp2ar0 commented on src/main/java/com/econo/racingGame/model/Rank.java:
```diff
@@ -0,0 +1,15 @@
+package com.econo.racingGame.model;
+
+public class Rank {
```
이 클래스는 어디서 쓰이는거죠? 찾아봤는데 저한테는 안보이네요..

SimKiSeong committed:
fix : 레이싱게임 step2 1차 수정
1. bigInteger함수 자바 라이브러리로 교채
2. 문자열 합치기 수정
3. makeCars 메소드 세분화
4. Rank클래스 삭제
5. 함수명 수정

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/GameController.java:
```diff
@@ -0,0 +1,50 @@
+public class GameController {
+
+    private final static int MOVE_BOUNDARY = 4;
+    private final static int RANDOM_BOUNDART = 10;
```
오타

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/GameController.java:
```diff
@@ -0,0 +1,50 @@
+        return random.nextInt(RANDOM_BOUNDART);
+    }
+
+    public int distance(int tryNumber) {
```
메소드 네이밍 규칙에 어긋납니다

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/GameController.java:
```diff
@@ -0,0 +1,50 @@
+        return cars;
+    }
+
+    public int randomNumber() {
```
메소드 네이밍 규칙에 어긋납니다

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/GameController.java:
```diff
@@ -0,0 +1,50 @@
+    public Car[] settingCars(String[] names, int tryNumber) {
+        for (int i = 0; i < cars.length; i++) {
+            cars[i].setDriver(names[i]);
+            cars[i].setMoveDistance(distance(tryNumber));
```
돌아가는 로직에는 문제가 없습니다.
하지만 `레이싱게임`의 관점으로 바라봤으면 좋겠습니다.
Car 하나하나마다 거리를 세팅해준다..?
객체지향적으로 코딩해보세요
'상태값을 가지고 있는 객체에 일을 시켜라'

hyukjin-lee replied:
맞아요 객체지향 생활체조 공부하고 모르는건 질문하라니까

esp2ar0 commented:
객체지향 생활체조에 나온 내용들을 다 반영해서 리팩토링하고, 자바 8 stream, lambda를 응용해서 반복문 최소화 해보세요

wijehyeon commented:
기능구현에는 문제가 없어보입니다. 잘하시네요
자잘자잘한 오타, 그리고 기본적인 방법들 말고 자바8의 기능들을 활용해보면 좋을 것 같습니다.

Bellroute commented on src/main/java/com/econo/racingGame/controller/GameController.java:
```diff
@@ -0,0 +1,50 @@
+
+    public void makeCars(int carNumbers) {
+        cars = new Car[carNumbers];
+        for (int i = 0;i<carNumbers;i++) {
```
ctrl + alt + L 을 눌러 코드 정리를 습관화해주세요

Bellroute commented on src/main/java/com/econo/racingGame/controller/GameController.java:
```diff
@@ -0,0 +1,50 @@
+    public int distance(int tryNumber) {
+        int distance = 0;
+        for (int i = 0; i < tryNumber; i++) {
+            distance = distance + checkRandomNumber(randomNumber());
```
distance += checkRandomNumber(randonNumber()); 로 코드를 줄일 수 있을 것 같습니다.

hyukjin-lee commented on src/main/java/com/econo/racingGame/controller/RankController.java:
```diff
@@ -0,0 +1,34 @@
+        return NONE;
+    }
+
+    public String[] winner(Car[] cars) {
```
메소드 네이밍에 계속 명사가 보이네

hyukjin-lee commented on src/main/java/com/econo/racingGame/view/InputView.java:
```diff
@@ -18,4 +20,8 @@ public static int tryNumber() {
-}
+    public static String[] divideDriver(String inputName) {
+        return inputName.split(SPLIT_MARK);
+    }
```
이런건 View에서 하지말고 util 패키지를 만들어서 StringUtils 클래스를 만들어 활용해보세요. 재사용가능하게

hyukjin-lee commented on src/main/java/com/econo/racingGame/view/ResultView.java:
```diff
@@ -4,19 +4,31 @@
-    private static final char BAR = '-';
+    private static final String BAR = "-";
+    private static final String RESULT_FORM = " : ";
+    private static final String WINNER_FORM = ", ";
```
이런것까지 하드코딩 안하려고 하다니... 스터디 규칙 지키려고 한 점 좋습니다 👍  근데 실전에선 이정도까진 안해도 되요 ㅎㅎ;

hyukjin-lee commented on src/test/java/GameControllerTest.java:
```diff
@@ -0,0 +1,52 @@
+
+import static junit.framework.TestCase.assertEquals;
+
+public class GameControllerTest {
```
테스트는 클래스별/메소드 테스트로 해주세요. '클래스명test'로 클래스 만들고 하시면 됩니다. 테스트할 수 있는 메소드들만 테스트하셔도 돼요

hyukjin-lee reviewed (changes requested):
저번에 리뷰 엄청 많이 받았던거 반영 안된게 보이네요. 전체적으로 객체지향적으로 짜면서 재사용성에 대해 생각하고 책임과 역할의 분리에 집중하면서 구현해주세요. 지금은 그냥 리뷰 보고 별다른 생각없이 반영한 것 같아서 아쉽습니다. 앞으로 리뷰보고 이해안되면 바로바로 말씀해주세요. 그리고 배열보다 웬만하면 Collection을 사용하시고 for문 if문들 자바8  stream lambda를 최대한 활용해서 리팩토링 해보세요

SimKiSeong committed:
fix : 자동차 움직임 변경
GameController -> Car

SimKiSeong committed:
fix : 2차 수정
1. move함수 자동차 내부로 이동
2. string, random generator제작
3. array 대신 list사용
4. 포문 람다형식으로 변

SimKiSeong committed:
add : 각 클래스별 테스트 제작 및 확인

hyukjin-lee commented on src/main/java/com/econo/racingGame/model/Car.java:
```diff
@@ -2,7 +2,28 @@
+        this.moveDistance = 0;
+    }
+
+    public Car(String driver,int moveDistance) {
```
라인정리 생활화

hyukjin-lee commented on src/main/java/com/econo/racingGame/model/Car.java:
```diff
@@ -12,4 +33,10 @@ public void setMoveDistance(int moveDistance) {
     }
 
-}
+    public void accelerate(boolean goFoward) {
```
```suggestion
    public void accelerate(boolean isForward) {
```

hyukjin-lee reviewed (approved):
멋집니다 💯 수고많이하셨어요 ! 멋진 코드네요

SimKiSeong committed:
fix : 라인 정리 및 변수명 수정

wijehyeon commented:
다져나가다 보니 이제 리뷰달게 없습니다

esp2ar0 commented on src/main/java/com/econo/racingGame/controller/RankController.java:
```diff
@@ -0,0 +1,36 @@
+        return StringGenerator.divideComma(winner);
+    }
+
+    public int farDistance(List<Car> cars) {
```
메소드는 동사
변수는 명사인데
두개의 이름이 같다는건 어느 한쪽이 잘못 네이밍 됐다는 거겠죠?

esp2ar0 commented:
스트림 사용 안하셨네요 ??
